crates/ide_assists/src/handlers/merge_match_arms.rs

```diff
@@ -88,6 +94,32 @@ fn contains_placeholder(a: &ast::MatchArm) -> bool {
     matches!(a.pat(), Some(ast::Pat::WildcardPat(..)))
 }
 
+fn are_same_types(
+    current_arm_types: &Vec<Option<hir::TypeInfo>>,
+    arm: &ast::MatchArm,
+    ctx: &AssistContext,
+) -> bool {
+    let arm_types = get_arm_types(&ctx, &arm);
+    for i in 0..arm_types.len() {
+        let other_arm_type = &arm_types[i];
+        let current_arm_type = &current_arm_types[i];
+        if let (Some(other_arm_type), Some(current_arm_type)) = (other_arm_type, current_arm_type) {
+            return &other_arm_type.original == &current_arm_type.original;
+        }
+    }
+
+    return true;
+}
+
+fn get_arm_types(ctx: &AssistContext, arm: &ast::MatchArm) -> Vec<Option<hir::TypeInfo>> {
+    match arm.pat() {
+        Some(ast::Pat::TupleStructPat(tp)) => {
+            tp.fields().into_iter().map(|field| ctx.sema.type_of_pat(&field)).collect_vec()
+        }
+        _ => Vec::new(),
```

Comment:
This will allocate a new vector (I presume) in the case that the type is not `TupleStructPat`. Aside from the assumption that it will always be `TupleStructPat`, this should be called rarely in the first place: the `same_text` requirement is done before this code gets executed. 

I looked at removing this potential allocation altogether by returning `Option<T>` but the code got quite a bit messier as a result so I decided on this instead. Let me know if you'd like to see this differently

Reply:
In Rust, empty `Vec`s do not allocate.

Reply:
Oh great! That's the best of both worlds then, thanks